### PR TITLE
Merge Main to feature 022723

### PR DESCRIPTION
**Pull request by charleshuang80**
**Summary:** Merge Main to Feature

Addresses [CUMULUS-XX: Develop amazing new feature](https://bugs.earthdata.nasa.gov/browse/CUMULUS-XXX)

## Changes

* Detailed list or prose of changes
* ...

## PR Checklist

- [ ] Update CHANGELOG
- [ ] Unit tests
- [ ] Ad-hoc testing - Deploy changes and test manually
- [ ] Integration tests

### CODE REVIEW

File: packages/api/lib/writeRecords/write-granules.js [hunk 1]
```diff
@@ -462,52 +511,95 @@ const _writeGranuleRecords = async (params) => {
...
+        // values are propagated to Dynamo and ES.
+        // The only values allowed to be updated in the PG granule write under these conditions are
+        // currently status, timestamp, updated_at, and created_at, and the associated execution
+        // as part of the write chain
+        const limitedUpdateApiGranuleRecord = await translatePostgresGranuleToApiGranule({
+          granulePgRecord: pgGranule,
+          knexOrTransaction: trx,
+        });
+
+        await granuleModel.storeGranule(limitedUpdateApiGranuleRecord, writeConstraints);
```
Comment (Jkovarik) [inline packages/api/lib/writeRecords/write-granules.js:548]: I'm concerned this is still here, given https://github.com/nasa/cumulus/pull/2954

We need to look into how we even got back to writing dynamo granules.

**Review by Jkovarik — CHANGES_REQUESTED** [review]
Requesting changes re: write-granules.js conversation

File: packages/api/lib/writeRecords/write-granules.js (hunk 1 above)
Comment (Jkovarik) [reply packages/api/lib/writeRecords/write-granules.js:548]: The above was influenced by the weird diff here - the feature branch at https://github.com/nasa/cumulus/blob/feature/rds-phase-3/packages/api/lib/writeRecords/write-granules.js#L372 doesn't have 

```
await granuleModel.storeGranule(apiGranuleRecord, writeConstraints);
```

We need to update this PR to *not* re-add dynamo writes.

File: packages/api/lib/schemas.js [hunk 2]
```diff
@@ -0,0 +1,753 @@
+'use strict';
+
+module.exports.accessToken = {
```
Comment (Jkovarik) [inline packages/api/lib/schemas.js]: Weird, why is this showing up as an addition, we'd already started using this in feature 🤔

File: packages/s3-credentials-endpoint/tests/test-handler-with-edl-token.js [hunk 3]
```diff
@@ -77,6 +74,13 @@ test.before((t) => {
   nock.enableNetConnect('localhost:80');
 });
 
+test.beforeEach((t) => {
```
Comment (Jkovarik) [inline packages/s3-credentials-endpoint/tests/test-handler-with-edl-token.js:77]: 👍🏻

File: packages/api/tests/lib/writeRecords/test-write-granules.js [hunk 4]
```diff
@@ -1662,7 +1652,6 @@ test.serial('writeGranulesFromMessage() removes preexisting granule file from Po
     granule,
     granulePgModel,
     knex,
-    providerCumulusId,
```
Comment (Jkovarik) [inline packages/api/tests/lib/writeRecords/test-write-granules.js:1665]: Where did this change come from?

Comment (Jkovarik) [reply packages/api/tests/lib/writeRecords/test-write-granules.js:1665]: It's a good change, but it looks like a refactor from the merge branch to eliminate a signature mismatch between the test and the call?     This seems fine, assuming it's not some other git ghost, although nit:there are other references in the file.

File: packages/api/tests/lib/writeRecords/test-write-granules.js [hunk 5]
```diff
@@ -2316,83 +3389,122 @@ test.serial('writeGranulesFromMessage() honors granule.createdAt time if provide
...
+    granule_id: granuleId,
+    collection_cumulus_id: collectionCumulusId,
+  });
+  const apiFormattedInitialPostgresGranule =
+    await translatePostgresGranuleToApiGranule({
+      granulePgRecord: initialPostgresRecord,
+      knexOrTransaction: knex,
+    });
 
-test.serial('writeGranulesFromMessage() falls back to workflow_start_time if granule.createdAt is not provided in cumulus_message for a granule', async (t) => {
```
Comment (Jkovarik) [inline packages/api/tests/lib/writeRecords/test-write-granules.js:2346]: Oof, this diff is rough - confirmed this is just a down-file move.

File: packages/api/tests/lib/writeRecords/test-write-granules.js [hunk 6]
```diff
@@ -2496,7 +3618,7 @@ test.serial('writeGranuleFromApi() removes preexisting granule file from postgre
   });
   const returnedGranule = await granulePgModel.create(knex, pgGranule, '*');
 
-  const [fakeFile] = await filePgModel.create(knex, {
+  const fakeFile = await filePgModel.create(knex, {
```
Comment (Jkovarik) [inline packages/api/tests/lib/writeRecords/test-write-granules.js:3621]: This doesn't seem correct - `.create` returns an array of type objects.

File: packages/api/tests/lib/writeRecords/test-write-granules.js [hunk 7]
```diff
@@ -2618,7 +3739,7 @@ test.serial('writeGranuleFromApi() writes a granule to PostgreSQL and Elasticsea
   );
 });
 
-test.serial('writeGranuleFromApi() writes a granule to PostgreSQL, DynamoDB, and Elasticsearch and populates a consistent createdAt default value', async (t) => {
```
Comment (Jkovarik) [inline packages/api/tests/lib/writeRecords/test-write-granules.js:2621]: 👍🏻  good catch

File: packages/api/tests/lib/writeRecords/test-write-granules.js [hunk 8]
```diff
@@ -2915,20 +4035,12 @@ test.serial('writeGranuleFromApi() given a granule with all fields populated is
     }
   );
 
-  // Validate that none of the responses come back as 'null', we want them removed, not set
```
Comment (Jkovarik) [inline packages/api/tests/lib/writeRecords/test-write-granules.js:2918]: Why was this removed?    It's present in main/feature, and seems like a valid assertion given a quick glance.

File: packages/api/tests/lib/writeRecords/test-write-granules.js [hunk 9]
```diff
@@ -1088,17 +1088,7 @@ test.serial('writeGranulesFromMessage() on re-write saves granule records to Dyn
     apiFormattedPostgresGranule,
     expectedGranule
   );
-  // ES fails here because of bad write logic in <TICKET>
```
Comment (Jkovarik) [inline packages/api/tests/lib/writeRecords/test-write-granules.js:1091]: Confirmed this was part of https://github.com/nasa/cumulus/pull/3242/files#diff-e4b95ee0e3b92433a8e670895c1e8da32d083c3ca64e31761d4bb353cbb791a7L1123 --- commenting here for posterity/future us.

**Review by Jkovarik — CHANGES_REQUESTED** [review]
@charleshuang80  thanks for addressing my original feedback. 

I did a manual main vs feature review, with particular attention to the changes in 3071/3111 to ensure the updates make it in - I have a couple of comments re: the units that should be looked at, but it's nearly there.  Nice work!

File: packages/api/tests/lib/writeRecords/test-write-granules.js (hunk 4 above)
Comment (charleshuang80) [reply packages/api/tests/lib/writeRecords/test-write-granules.js:1665]: `providerCumulusId` seems to be passed in in the feature branch. I did a manual removal of a lot of them, but I think some of them came back when I resolved the merge conflicts and not sure exactly why that is. It was taken out in main. Not sure if I should go back through the whole file again and manually take them out.

File: packages/api/tests/lib/writeRecords/test-write-granules.js (hunk 5 above)
Comment (charleshuang80) [reply packages/api/tests/lib/writeRecords/test-write-granules.js:2346]: yeah

File: packages/api/tests/lib/writeRecords/test-write-granules.js (hunk 6 above)
Comment (charleshuang80) [reply packages/api/tests/lib/writeRecords/test-write-granules.js:3621]: hmm, I can change this back. I did not make this change, but can try and investigate when/how it got in if helpful.

File: packages/api/tests/lib/writeRecords/test-write-granules.js (hunk 8 above)
Comment (charleshuang80) [reply packages/api/tests/lib/writeRecords/test-write-granules.js:2918]: It is a duplicated assertion - see line 2944/4056. I was not sure which one to keep, opted for the latter one (let me know if that should change).

Comment (Jkovarik) [reply packages/api/tests/lib/writeRecords/test-write-granules.js:2918]: Ah - got it, that totally makes sense.

File: packages/api/lib/writeRecords/write-granules.js (hunk 1 above)
Comment (Jkovarik) [reply packages/api/lib/writeRecords/write-granules.js]: This was addressed with the commits following the comment.

**Review by Jkovarik — APPROVED** [review]
These updates look good - assuming there aren't significant changes/fixes missed (in which case please ask for a re-review), this looks good.
